### DISCUSSION TITLE
v8: How to derive the Cell render type?

### DISCUSSION THREAD

**Discussion by Obiwarn** [TanStack/table#3823] [Q&A] [answered]
Is there a way to get the type of the finally rendered cell?

So that I could do something like this (e.g. conditionally formatting numbers or plain strings):

```
  {row.getVisibleCells().map((cell) => {
    const cellContent = cell.renderCell();
    const isNumberCell = typeof cellContent === "number";
      // ...
```

**Comment by tannerlinsley** [comment, marked as answer]
I think cell.value is what you’re looking for :)

**Reply by Obiwarn** [reply]
Not really. I am looking for the cell render function (see example below). 

What I try to accomplish is a kind of "default cell":
If the cell function just returns a string (like in the example), or there is no cell function at all, then surround it with a div and some padding, for instance. If it returns a proper React component leave it to the user. 

```
table.createDataColumn((o) => o.metadata.filingDate.value, {
            id: "filingDate",
            cell: ({ value }) => formatDate(value),
          }),
```

**Reply by colindjk** [reply]
I haven't tested it, but I think if use the `render` option for `useTable`, you can do something like this 

```tsx
// add whatever default rendering you'd like
const defaultRender = (val) => <div style={{padding: '10px'}}>{val}</div>;

export const render = <TProps extends {}>(
  Comp: Renderable<TProps>,
  props: TProps
): React.ReactNode =>
  !Comp ? null : isReactComponent(Comp) ? <Comp {...props} /> : defaultRender(Comp)

function isReactComponent(component: unknown): component is React.FC {
  return (
    isClassComponent(component) ||
    typeof component === 'function' ||
    isExoticComponent(component)
  )
}
```

Most of the above code was pulled from [react-tables code here](https://github.com/TanStack/react-table/blob/8976e792c9a900b25cf93fb095b83325d6ee38ab/packages/react-table/src/index.tsx#L20).

**Reply by tannerlinsley** [reply]
Do not do the above. Just add a default cell option like this: 

```tsx
const defaultColumn = {
  cell: (props) => <div style={{padding: '10px'}}>{props.value}</div>
}

// ...

const tableInstance = useTable({
  defaultColumn
})
```

**Reply by Obiwarn** [reply]
Thank you! This works.

However, I get typing errors for the `cell` function:

`TS2322: Type '(props: { instance: TableInstance<T>; row: Row<T>; column: Column<T>; cell: Cell<T>; value: T["Value"]; }) => Element' is not assignable to type 'Renderable<T, { instance: TableInstance<T>; row: Row<T>; column: Column<T>; cell: Cell<T>; value: T["Value"]; }> | undefined'.   Type '(props: { instance: TableInstance<T>; row: Row<T>; column: Column<T>; cell: Cell<T>; value: T["Value"]; }) => Element' is not assignable to type '(props: { instance: TableInstance<T>; row: Row<T>; column: Column<T>; cell: Cell<T>; value: T["Value"]; }) => ReturnType<UseRenderer<T>>'.     Type 'Element' is not assignable to type 'ReturnType<UseRenderer<T>>'. `

The code:

```
interface Props<T extends AnyGenerics> {
  table: Table<T>;
}

export const SuperTable = <T extends AnyGenerics>(props: Props<T>) => { 
  type TableGenerics = typeof props.table.generics

  const defaultColumn: Partial<ColumnDef<TableGenerics>> = {
    cell: (props) => <div style={{ padding: "40px" }}>{props.value}</div>,   // <-- here error
  };

// ...
```

**Reply by tannerlinsley** [reply]
If you have no idea what your generics are, you can just use the exported `TableGenerics` type instead of your own generic variable. If you *do* know your generics, this is probably what you want:

```tsx
let table = createTable()

type TableGenerics = typeof table.generics

interface Props {
  table: Table<TableGenerics>
}

export const SuperTable = (props: Props) => {
  const defaultColumn: Partial<ColumnDef<TableGenerics>> = {
    cell: props => <div style={{ padding: '40px' }}>{props.value}</div>,
  }

  const instance = useTableInstance(table, {
    columns: [],
    data: [],
    defaultColumn,
    getCoreRowModel: null!
  })
}
```

**Reply by Obiwarn** [reply]
@tannerlinsley when using the exported `TableGenerics` I get some typing errors (I guess that is a bug?) :

[Codesandbox](https://codesandbox.io/s/eloquent-roentgen-442vu6?file=/src/test.tsx)

I use    "@tanstack/react-table": "^8.0.0-alpha.58",

```

import {
  ColumnDef,
  createTable,
  Table,
  TableGenerics,
  useTableInstance,
} from "@tanstack/react-table";
import React from "react";

let table = createTable();

interface Props {
  table: Table<TableGenerics>;
}

export const Test = () => <SuperTable table={table}></SuperTable>;

const SuperTable = (props: Props) => {
  const defaultColumn: Partial<ColumnDef<TableGenerics>> = {
    cell: (props) => (
      <div style={{ padding: "40px" }}>{props.value as any}</div>
    ),
  };

  const instance = useTableInstance(table, {
    columns: [],
    data: [],
    defaultColumn,
    getCoreRowModel: null!,
  });
  return <div></div>;
};

```
